### ISSUE TITLE
Add ability to define custom predicate functions

### ISSUE THREAD

**Issue by neetle** [expr-lang/expr#497]
Hi! 

I'm currently working on a project that would like to use this as the primary language to build filters for it's frontend. 

At the moment, we're hoping to allow users to write queries like `Nodes | filter(.Name=="cat")`. We're currently facing an issue though, as `Nodes` isn't a slice. This leads to issues like `array.Len()` causing panics in the VM every time we hit an `OpBegin`. 

As we don't control the layout/go type of the data we're passing into `expr`, would you be against defining some way to implement functions that take custom predicates? 

Looking forwards, this could also be a good place to look into support for the changes proposed in https://github.com/golang/go/issues/61898.

**Comment by antonmedv** [comment]
Yes, it is possible to implement next expression, even if `Nodes` is not an array.

    Nodes | filter(.Name=="cat")

To do it you can create a [patcher](https://expr-lang.org/docs/Visitor-and-Patch) which will be wrapping Nodes inside filter into a helper function. The helper function will transform Nodes to an array.

After the patcher resulting code should look like this:

   nodesToArray(Nodes) | filter(.Name=="cat")
